### PR TITLE
feat: Restructure with site branch config relation

### PR DESCRIPTION
**Pull request by apburnes**
Apart of #4101 
Closes https://github.com/cloud-gov/pages-core/issues/4175

## Changes proposed in this pull request:
- [x] Add unique index on `branch` and `siteId` to table
- [x] Add `s3Key` to site branch config table
- [x] Add `context` to site branch config table
- [x] Add site branch config API endpoints
- [x] Update site settings UI to work with new branch config
- [x] Move automatic build kickoff from site update to site branch config creation or update.
- [x] Add custom domains to its own page under a site

### Custom Domains Page
Screenshot 2023-07-21 at 11 35 00 AM

### Site Branch Config
https://github.com/cloud-gov/pages-core/assets/693815/c4043c61-c49b-429d-8446-2b18f3ed814c

## security considerations
None.

### CODE REVIEW

File: frontend/components/site/SiteSettings/BranchConfig.jsx [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                      });
+                    }}
+                  />
+                </div>
+              )}
+              <div>
+                {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
+                <label htmlFor={`${branch}-config-input`}>Configuration:</label>
+                <p>
+                  Add additional configuration in yaml to be added to your
```
Comment (svenaas) [inline frontend/components/site/SiteSettings/BranchConfig.jsx:80]:
```suggestion
                  Add additional configuration in YAML to be added to your
```

Comment (svenaas) [reply frontend/components/site/SiteSettings/BranchConfig.jsx:80]: (For consistency with the error messages we raise from our validators.)

File: frontend/components/site/SiteSettings/index.jsx [hunk 2]
```diff
@@ -35,27 +38,15 @@ function SiteSettings() {
...
     engine: site.engine,
-    defaultConfig: site.defaultConfig || '',
-    demoConfig: site.demoConfig || '',
-    previewConfig: site.previewConfig || '',
   };
 
   return (
     <div>
       <p>
         See our documentation site for more about
```
Comment (svenaas) [inline frontend/components/site/SiteSettings/index.jsx:48]: Should this be a link?

File: frontend/components/site/SiteSettings/BranchConfigs.jsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      .catch(error => setConfigs({ ...configs, state: 'error', error: error.message }));
+  }, [siteId]);
+
+  return (
+    <div>
+      <h3>Branch Configurations</h3>
+      {configs.error && (
+        <div className="well">
+          <h4>
+            An error occured while loading you site branch configurations.
```
Comment (svenaas) [inline frontend/components/site/SiteSettings/BranchConfigs.jsx:45]:
```suggestion
            An error occurred while loading your site branch configurations.
```

File: api/controllers/site-branch-config.js [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+          siteBranchConfig: { id: siteBranchConfig.id, siteId },
+        }
+      );
+
+      const json = serialize(siteBranchConfig);
+
+      return res.ok(json);
+    } catch (error) {
+      return res.badRequest({
+        message: `An error occurred updated the site branch config: ${error.message}`,
```
Comment (svenaas) [inline api/controllers/site-branch-config.js:211]:
```suggestion
        message: `An error occurred updating the site branch config: ${error.message}`,
```

File: api/controllers/site-branch-config.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+  }
+
+  return `preview/${site.owner}/${site.repository}/${branch}`;
+}
+
+function validate({ branch, config = {}, context } = {}) {
+  const parsedConfig = parseSiteConfig(config);
+
+  if (context && typeof context !== 'string') {
+    throw new ValidationError('Context must be valid string.');
```
Comment (svenaas) [inline api/controllers/site-branch-config.js:31]:
```suggestion
    throw new ValidationError('Context must be a valid string.');
```

**Review by svenaas — CHANGES_REQUESTED** [review]
A few small copy suggestions.

File: api/models/site-branch-config.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+          if (this.context !== 'preview' && !this.branch) {
+            throw new Error(
+              'Branch attribute cannot be null when context attribute is not preview'
+            );
+          }
+        },
+        nonPreviewS3KeyDefined() {
+          if (this.context !== 'preview' && !this.s3Key) {
+            throw new Error(
+              'The s3Key attribute cannot be null when context attribute is not preview'
```
Comment (drewbo) [inline api/models/site-branch-config.js:68]: I'm getting this error when trying to edit the live site config on a test site in dev

**Review by drewbo — DISMISSED** [review]
I had one small question about flow and I'm seeing an error on `s3Key` validation when trying to edit a live site config. Once it's ready we should queue up a docs PR to match. Happy to pair review this later

File: api/controllers/site.js [hunk 7]
```diff
@@ -198,22 +223,23 @@ module.exports = wrapHandlers({
...
+    }
+
+    if (site.defaultBranch) {
+      await Build.create({
+        user: req.user.id,
+        site: site.id,
+        branch: site.defaultBranch,
+        username: req.user.username,
+      }).then(b => b.enqueue());
     }
```
[the remark this answers is not in the document]
Comment (apburnes) [reply api/controllers/site.js:243]: The logic didn't work to before and this probably needs to be ripped out. Before, any update to a live site or demo site config would kick off two builds.

File: api/models/site-branch-config.js (hunk 6 above)
Comment (apburnes) [reply api/models/site-branch-config.js:68]: This is probably due to the migration with the updated commits post migration. We may need to run the migration back and forward against the dev database manually.

Comment (apburnes) [reply api/models/site-branch-config.js:68]: Fixed with update to migration script.

File: frontend/components/site/SiteSettings/index.jsx (hunk 2 above)
Comment (apburnes) [reply frontend/components/site/SiteSettings/index.jsx:48]: It's linked on the next few lines.

**Review by drewbo — APPROVED** [review]

**Pull request merged**